File: apollo-runtime/src/test/java/com/apollographql/apollo/response/ScalarTypeAdaptersTest.java
package com.apollographql.apollo.response;

import com.apollographql.apollo.api.CustomTypeAdapter;
import com.apollographql.apollo.api.CustomTypeValue;
import com.apollographql.apollo.api.ScalarType;
import com.apollographql.apollo.api.ScalarTypeAdapters;
import com.apollographql.apollo.api.internal.UnmodifiableMapBuilder;
import org.junit.Test;
import org.mockito.Mockito;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static com.google.common.truth.Truth.assertThat;

public class ScalarTypeAdaptersTest {

  @Test
  public void customAdapterTakePrecedentOverDefault() {
    final Map<ScalarType, CustomTypeAdapter<?>> customTypeAdapters = new HashMap<>();
    final CustomTypeAdapter expectedAdapter = Mockito.mock(CustomTypeAdapter.class);
    customTypeAdapters.put(new ScalarType() {
      @Override
      public String typeName() {
        return "String";
      }

      @Override
      public String className() {
        return String.class.getName();
      }
    }, expectedAdapter);

    final CustomTypeAdapter actualAdapter = new ScalarTypeAdapters(customTypeAdapters).adapterFor(new ScalarType() {
      @Override
      public String typeName() {
        return "String";
      }

      @Override
      public String className() {
        return String.class.getName();
      }
    });

    assertThat(actualAdapter).isEqualTo(expectedAdapter);
  }

  @Test(expected = IllegalArgumentException.class)
  public void missingAdapter() {
    new ScalarTypeAdapters(Collections.<ScalarType, CustomTypeAdapter<?>>emptyMap())
        .adapterFor(
            new ScalarType() {
              @Override
              public String typeName() {
                return "RuntimeException";
              }

              @Override
              public String className() {
                return RuntimeException.class.getName();
              }
            }
        );
  }

  @Test
  public void defaultStringAdapter() {
    final CustomTypeAdapter<String> adapter = defaultAdapter(String.class);
    assertThat(adapter.decode(CustomTypeValue.fromRawValue("string"))).isEqualTo("string");
    assertThat(adapter.encode("string").value).isEqualTo("string");
  }

  @Test
  public void defaultBooleanAdapter() {
    final CustomTypeAdapter<Boolean> adapter = defaultAdapter(Boolean.class);
    assertThat(adapter.decode(CustomTypeValue.fromRawValue(true))).isEqualTo(true);
    assertThat(adapter.encode(true).value).isEqualTo(true);
  }

  @Test
  public void defaultIntegerAdapter() {
    final CustomTypeAdapter<Integer> adapter = defaultAdapter(Integer.class);
    assertThat(adapter.decode(CustomTypeValue.fromRawValue(100))).isEqualTo(100);
    assertThat(adapter.encode(100).value).isEqualTo(100);
  }

  @Test
  public void defaultLongAdapter() {
    final CustomTypeAdapter<Long> adapter = defaultAdapter(Long.class);
    assertThat(adapter.decode(CustomTypeValue.fromRawValue(100L))).isEqualTo(100L);
    assertThat(adapter.encode(100L).value).isEqualTo(100L);
  }

  @Test
  public void defaultFloatAdapter() {
    final CustomTypeAdapter<Float> adapter = defaultAdapter(Float.class);
    assertThat(adapter.decode(CustomTypeValue.fromRawValue(10.10f))).isWithin(0.0f).of(10.10f);
    assertThat(adapter.encode(10.10f).value).isEqualTo(10.10f);
  }

  @Test
  public void defaultDoubleAdapter() {
    final CustomTypeAdapter<Double> adapter = defaultAdapter(Double.class);
    assertThat(adapter.decode(CustomTypeValue.fromRawValue(10.10d))).isWithin(0.0d).of(10.10d);
    assertThat(adapter.encode(10.10d).value).isEqualTo(10.10d);
  }

  @Test
  public void defaultObjectAdapter() {
    final CustomTypeAdapter<Object> adapter = defaultAdapter(Object.class);
    assertThat(adapter.decode(CustomTypeValue.fromRawValue(RuntimeException.class))).isEqualTo("class java.lang.RuntimeException");
    assertThat(adapter.encode(RuntimeException.class).value).isEqualTo("class java.lang.RuntimeException");
  }

  @Test
  public void defaultMapAdapter() {
    final Map<String, Object> value = new UnmodifiableMapBuilder<String, Object>()
        .put("key1", "value1")
        .put("key2", "value2")
        .build();

    final CustomTypeAdapter<Map> adapter = defaultAdapter(Map.class);
    assertThat(adapter.decode(CustomTypeValue.fromRawValue(value))).isEqualTo(value);
    assertThat(adapter.encode(value).value).isEqualTo(value);
  }

  @Test
  public void defaultListAdapter() {
    final List<String> value = Arrays.asList("item 1", "item 2");
    final CustomTypeAdapter<List> adapter = defaultAdapter(List.class);
    assertThat(adapter.decode(CustomTypeValue.fromRawValue(value))).isEqualTo(value);
    assertThat(adapter.encode(value).value).isEqualTo(value);
  }

  @Test
  public void defaultJsonString() {
    final CustomTypeValue.GraphQLJsonObject actualObject = new CustomTypeValue.GraphQLJsonObject(
        new UnmodifiableMapBuilder<String, Object>()
            .put("key", "scalar")
            .put("object", new UnmodifiableMapBuilder<String, Object>()
                .put("nestedKey", "nestedScalar")
                .build()
            )
            .put("list", Arrays.asList("1", "2", "3"))
            .build()
    );
    final String expectedJsonString = "{\"key\":\"scalar\",\"object\":{\"nestedKey\":\"nestedScalar\"},\"list\":[\"1\",\"2\",\"3\"]}";

    final CustomTypeAdapter<String> adapter = defaultAdapter(String.class);
    assertThat(adapter.decode(actualObject)).isEqualTo(expectedJsonString);
  }

  private <T> CustomTypeAdapter<T> defaultAdapter(final Class<T> clazz) {
    return new ScalarTypeAdapters(Collections.<ScalarType, CustomTypeAdapter<?>>emptyMap()).adapterFor(
        new ScalarType() {
          @Override
          public String typeName() {
            return clazz.getSimpleName();
          }

          @Override
          public String className() {
            return clazz.getName();
          }
        }
    );
  }
}

File: apollo-runtime/src/test/java/com/apollographql/apollo/response/ScalarTypeAdaptersTest.kt
package com.apollographql.apollo.response

import com.apollographql.apollo.api.CustomScalarTypeAdapter
import com.apollographql.apollo.api.JsonElement
import com.apollographql.apollo.api.ScalarType
import com.apollographql.apollo.api.ScalarTypeAdapters
import com.google.common.truth.Truth.assertThat
import org.junit.Test

class ScalarTypeAdaptersTest {
  @Test
  fun customAdapterTakePrecedentOverDefault() {
    val customScalarTypeAdapters = mutableMapOf<ScalarType, CustomScalarTypeAdapter<*>>()
    val expectedAdapter = MockCustomScalarTypeAdapter()
    customScalarTypeAdapters[object : ScalarType {
      override val graphqlName = "String"
      override val className: String
        get() = String::class.java.name
    }] = expectedAdapter

    val actualAdapter = ScalarTypeAdapters(customScalarTypeAdapters).adapterFor<String>(object : ScalarType {
      override val graphqlName = "String"
      override val className: String
        get() = String::class.java.name
    })
    assertThat(actualAdapter).isEqualTo(expectedAdapter)
  }

  @Test(expected = IllegalArgumentException::class)
  fun missingAdapter() {
    ScalarTypeAdapters(emptyMap())
        .adapterFor<RuntimeException>(
            object : ScalarType {
              override val graphqlName = "RuntimeException"
              override val className: String
                get() = RuntimeException::class.java.name
            }
        )
  }

  @Test
  fun defaultStringAdapter() {
    val adapterScalar: CustomScalarTypeAdapter<String> = defaultAdapter(String::class.java)
    assertThat(adapterScalar.decode(JsonElement.fromRawValue("string"))).isEqualTo("string")
    assertThat(adapterScalar.encode("string").toRawValue()).isEqualTo("string")
  }

  @Test
  fun defaultBooleanAdapter() {
    val adapterScalar: CustomScalarTypeAdapter<Boolean> = defaultAdapter(Boolean::class.java)
    assertThat(adapterScalar.decode(JsonElement.fromRawValue(true))).isEqualTo(true)
    assertThat(adapterScalar.encode(true).toRawValue()).isEqualTo(true)
  }

  @Test
  fun defaultIntegerAdapter() {
    val adapterScalar: CustomScalarTypeAdapter<Int> = defaultAdapter(Int::class.java)
    assertThat(adapterScalar.decode(JsonElement.fromRawValue(100))).isEqualTo(100)
    assertThat(adapterScalar.encode(100).toRawValue()).isEqualTo(100)
  }

  @Test
  fun defaultLongAdapter() {
    val adapterScalar: CustomScalarTypeAdapter<Long> = defaultAdapter(Long::class.java)
    assertThat(adapterScalar.decode(JsonElement.fromRawValue(100L))).isEqualTo(100L)
    assertThat(adapterScalar.encode(100L).toRawValue()).isEqualTo(100L)
  }

  @Test
  fun defaultFloatAdapter() {
    val adapterScalar: CustomScalarTypeAdapter<Float> = defaultAdapter(Float::class.java)
    assertThat(adapterScalar.decode(JsonElement.fromRawValue(10.10f))).isWithin(0.0f).of(10.10f)
    assertThat(adapterScalar.encode(10.10f).toRawValue()).isEqualTo(10.10f)
  }

  @Test
  fun defaultDoubleAdapter() {
    val adapterScalar: CustomScalarTypeAdapter<Double> = defaultAdapter(Double::class.java)
    assertThat(adapterScalar.decode(JsonElement.fromRawValue(10.10))).isWithin(0.0).of(10.10)
    assertThat(adapterScalar.encode(10.10).toRawValue()).isEqualTo(10.10)
  }

  @Test
  fun defaultObjectAdapter() {
    val adapterScalar: CustomScalarTypeAdapter<Any> = defaultAdapter(Any::class.java)
    assertThat(adapterScalar.decode(JsonElement.fromRawValue(RuntimeException::class.java))).isEqualTo("class java.lang.RuntimeException")
    assertThat(adapterScalar.encode(RuntimeException::class.java).toRawValue()).isEqualTo("class java.lang.RuntimeException")
  }

  @Test
  fun defaultMapAdapter() {
    val value = mapOf<String, Any>(
        "key1" to "value1",
        "key2" to "value2"
    )
    val adapterScalar: CustomScalarTypeAdapter<Map<*, *>> = defaultAdapter(Map::class.java)
    assertThat(adapterScalar.decode(JsonElement.fromRawValue(value))).isEqualTo(value)
    assertThat(adapterScalar.encode(value).toRawValue()).isEqualTo(value)
  }

  @Test
  fun defaultListAdapter() {
    val value = listOf("item 1", "item 2")
    val adapterScalar: CustomScalarTypeAdapter<List<*>> = defaultAdapter(List::class.java)
    assertThat(adapterScalar.decode(JsonElement.fromRawValue(value))).isEqualTo(value)
    assertThat(adapterScalar.encode(value).toRawValue()).isEqualTo(value)
  }

  @Test
  fun defaultJsonString() {
    val actualObject = JsonElement.JsonObject(
        mapOf(
            "key" to JsonElement.JsonString("scalar"),
            "object" to JsonElement.JsonObject(mapOf("nestedKey" to JsonElement.JsonString("nestedScalar"))),
            "list" to JsonElement.JsonList(
                listOf(
                    JsonElement.JsonString("1"),
                    JsonElement.JsonString("2"),
                    JsonElement.JsonString("3"),
                )
            )
        )
    )
    val expectedJsonString = "{\"key\":\"scalar\",\"object\":{\"nestedKey\":\"nestedScalar\"},\"list\":[\"1\",\"2\",\"3\"]}"
    val adapter = defaultAdapter(String::class.java)
    assertThat(adapter.decode(actualObject)).isEqualTo(expectedJsonString)
  }

  private fun <T : Any> defaultAdapter(clazz: Class<T>): CustomScalarTypeAdapter<T> {
    return ScalarTypeAdapters(emptyMap()).adapterFor<T>(
        object : ScalarType {
          override val graphqlName: String
            get() = clazz.simpleName
          override val className: String
            get() = clazz.name
        }
    )
  }

  private inner class MockCustomScalarTypeAdapter : CustomScalarTypeAdapter<Any?> {
    override fun decode(jsonElement: JsonElement): Any {
      throw UnsupportedOperationException()
    }

    override fun encode(value: Any?): JsonElement {
      throw UnsupportedOperationException()
    }
  }
}